### PR TITLE
Add `RzConfig` for asm plugins

### PR DESCRIPTION
**Pull request by Rot127**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Adds an `RzConfig` to asm plugins and some callbacks to communicate with them.

It also updates the hexagon asm plugin to use the new plugins config.

It works roughly like that:

- On plugin `init()`: Plugin initializes its `RzConfig` struct and adds settings (see `hexagon_init(void **user)`).
- `RzAsm.get_config` callback returns the `RzConfig` struct from the plugin.
- On plugin load the config nodes are copied to the `RzCore` config.
- If a config is set via the shell, it sets the `RzCores` copy of the config node.
- The callback of this setting is called and the `RzCores` copy of the node is given to it.
- The cb should be implemented by the plugin!
  In the Hexagon case it simply copies the newly set settings from the `RzCore` node the plugins node.
  But in general the plugin can do with the information as it pleases.
- If the asm plugin was put out of use the `RzCores` copies of the config nodes are deleted.

**Test plan**

- The hexagon analysis tests cover the working of plugin settings.
- (un)set of plugins settings on asm plugin init was tested manually. 

**Closing issues**

None
Prerequisite for https://github.com/rizinorg/rizin/pull/2490

### CODE REVIEW

**Comment by Rot127** [comment]
Would also add an entry in Rizin Book. But currently missing the time for that.

**Comment by Rot127** [comment]
Also, if plugin settings are set in `./config/rizin/rizinrc` they won;t be applied because the plugin is not loaded yet.
I would like to address this bug after the next release. Because time and, I assume, most people won't need it yet.

File: librz/config/config.c [hunk 1]
```diff
@@ -352,6 +352,17 @@ RZ_API RzConfigNode *rz_config_set(RzConfig *cfg, RZ_NONNULL const char *name, c
...
 }
 
+RZ_API bool rz_config_add_node(RZ_BORROW RzConfig *cfg, RZ_OWN RzConfigNode *node) {
+	rz_return_val_if_fail(cfg && node, false);
+	if (!cfg->lock) {
+		ht_pp_insert(cfg->ht, node->name, node);
+		rz_list_append(cfg->nodes, node);
+		return true;
+	}
+	RZ_LOG_WARN("Config locked. Plugin config node not copied.\n");
```
Comment (thestr4ng3r) [inline librz/config/config.c]: Free the node here

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
Apart from that, lgtm. The config registration should maybe happen at plugin load time rather than `rz_asm_use()` but since the init/fini also happens there, I think this is ok for now.

File: librz/asm/asm.c [hunk 2]
```diff
@@ -304,13 +305,54 @@ RZ_API bool rz_asm_use_assembler(RzAsm *a, const char *name) {
...
+static void set_plugin_configs(RZ_BORROW RzAsm *rz_asm, RZ_BORROW RzConfig *pcfg) {
+	rz_return_if_fail(pcfg && rz_asm);
+
+	RzConfig *conf = ((RzCore *)(rz_asm->core))->config;
+	RzConfigNode *n;
+	RzListIter *it;
+	rz_list_foreach_iter(pcfg->nodes, it) {
+		n = it->data;
+		if (!rz_config_add_node(conf, rz_config_node_clone(n))) {
+			RZ_LOG_WARN("Failed add %s to the global config.\n", n->name)
```
Comment (XVilka) [inline librz/asm/asm.c]: I recommend putting the name in the quotes: `\"%s\"`
Also `failed add` sounds odd.

File: librz/asm/asm.c [hunk 3]
```diff
@@ -304,13 +305,54 @@ RZ_API bool rz_asm_use_assembler(RzAsm *a, const char *name) {
...
+static void unset_plugins_config(RZ_BORROW RzAsm *rz_asm, RZ_BORROW RzConfig *pcfg) {
+	rz_return_if_fail(pcfg && rz_asm);
+
+	RzConfig *conf = ((RzCore *)(rz_asm->core))->config;
+	RzConfigNode *n;
+	RzListIter *it;
+	rz_list_foreach_iter(pcfg->nodes, it) {
+		n = it->data;
+		if (!rz_config_rm(conf, n->name)) {
+			RZ_LOG_WARN("Failed remove %s from the global config.", n->name)
```
Comment (XVilka) [inline librz/asm/asm.c]: Same here

File: librz/asm/asm.c [hunk 4]
```diff
@@ -304,13 +305,54 @@ RZ_API bool rz_asm_use_assembler(RzAsm *a, const char *name) {
...
+/**
+ * \brief Deletes all copies of \p pcfg nodes in the RzConfig from \p rz_asm.
+ *
+ * \param rz_asm Pointer to RzAsm struct.
+ * \param pcfg Pointer to the plugins RzConfig struct.
+ */
+static void unset_plugins_config(RZ_BORROW RzAsm *rz_asm, RZ_BORROW RzConfig *pcfg) {
+	rz_return_if_fail(pcfg && rz_asm);
+
+	RzConfig *conf = ((RzCore *)(rz_asm->core))->config;
```
Comment (XVilka) [inline librz/asm/asm.c:337]: I don't like this kind of access. But I guess right now there is no better way.

File: librz/config/config.c [hunk 5]
```diff
@@ -352,6 +352,17 @@ RZ_API RzConfigNode *rz_config_set(RzConfig *cfg, RZ_NONNULL const char *name, c
 	return node;
 }
 
+RZ_API bool rz_config_add_node(RZ_BORROW RzConfig *cfg, RZ_OWN RzConfigNode *node) {
```
Comment (XVilka) [inline librz/config/config.c:362]: Add Doxygen please.

File: librz/config/config.c [hunk 6]
```diff
@@ -352,6 +352,17 @@ RZ_API RzConfigNode *rz_config_set(RzConfig *cfg, RZ_NONNULL const char *name, c
...
+	rz_return_val_if_fail(cfg && node, false);
+	if (!cfg->lock) {
```
Comment (XVilka) [inline librz/config/config.c]: Maybe reverse the condition?

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**